libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp

```diff
@@ -0,0 +1,182 @@
+#endif
+
+#ifndef RANGEFINDER_HEALTH_MAX
+ # define RANGEFINDER_HEALTH_MAX 3          // number of good reads that indicates a healthy rangefinder
+#endif
+
+void AP_SurfaceDistance::update()
+{
+    RangeFinder *rangefinder = RangeFinder::get_singleton();
+    if (rangefinder == nullptr) {
+        alt_healthy = false;
+        return;
+    }
+
+#if RANGEFINDER_TILT_CORRECTION == 1
+    const float tilt_correction = MAX(0.707f, AP::ahrs().get_rotation_body_to_ned().c.z);
+#else
+    const float tilt_correction = 1.0f;
+#endif
+
+    const uint32_t now = AP_HAL::millis();
+    // reset status and add to the bitmask as we progress through the update
+    status = 0;
+
+    // update enabled status
+    if (enabled) {
+        set_status(Status::Enabled);
+    }
+
+    // update health
+    alt_healthy = ((rangefinder->status_orient(rotation) == RangeFinder::Status::Good) &&
+                            (rangefinder->range_valid_count_orient(rotation) >= RANGEFINDER_HEALTH_MAX));
+    if (!alt_healthy){
+        set_status(Status::Unhealthy);
+    }
+
+    // tilt corrected but unfiltered, not glitch protected alt
+    alt_cm = tilt_correction * rangefinder->distance_cm_orient(rotation);
+
+    // remember inertial alt to allow us to interpolate rangefinder
+    inertial_alt_cm = inertial_nav.get_position_z_up_cm();
+
+    // glitch handling.  rangefinder readings more than RANGEFINDER_GLITCH_ALT_CM from the last good reading
+    // are considered a glitch and glitch_count becomes non-zero
+    // glitches clear after RANGEFINDER_GLITCH_NUM_SAMPLES samples in a row.
+    // glitch_cleared_ms is set so surface tracking (or other consumers) can trigger a target reset
+    const int32_t glitch_cm = alt_cm - alt_cm_glitch_protected;
+    bool reset_terrain_offset = false;
+    if (glitch_cm >= RANGEFINDER_GLITCH_ALT_CM) {
+        glitch_count = MAX(glitch_count+1, 1);
+        set_status(Status::Glitch_Detected);
+    } else if (glitch_cm <= -RANGEFINDER_GLITCH_ALT_CM) {
+        glitch_count = MIN(glitch_count-1, -1);
+        set_status(Status::Glitch_Detected);
+    } else {
+        glitch_count = 0;
+        alt_cm_glitch_protected = alt_cm;
+    }
+    if (abs(glitch_count) >= RANGEFINDER_GLITCH_NUM_SAMPLES) {
+        // clear glitch and record time so consumers (i.e. surface tracking) can reset their target altitudes
+        glitch_count = 0;
+        alt_cm_glitch_protected = alt_cm;
+        glitch_cleared_ms = now;
+        reset_terrain_offset = true;
+    }
+
+    // filter rangefinder altitude
+    const bool timed_out = now - last_healthy_ms > RANGEFINDER_TIMEOUT_MS;
+    if (alt_healthy) {
+        if (timed_out) {
+            // reset filter if we haven't used it within the last second
+            alt_cm_filt.reset(alt_cm);
+            reset_terrain_offset = true;
+            set_status(Status::Stale_Data);
+        } else {
+            alt_cm_filt.apply(alt_cm, 0.05f);
```

Comment:
Not necessarily a issue here, but this is assuming a 20Hz update rate. Excitingly plane runs its rangefinders at 50Hz....

Reply:
ok, did not know about that difference in rangefinder, but that can come when we work to integrate this into plane.

Reply:
I have made this a more generic now, by getting last loop time from AP::scheduler.  This **should** be a negligible change to copter behaviour.

Reply:
These do not run at loop rate, there running slower in the copter scheduler table. The 0.05 was OK, we just need to remember it when we port to another vehicle.